### ISSUE TITLE
PDE incorrectly removes final line break on sketch re-open

### ISSUE THREAD

**Issue by jeremydouglass** [processing/processing#5327]
## Description

Opening a Processing sketch in PDE loads the text with the final linebreak missing. Saving any tab then introduces an additional change that strips away the final line break on that tab.

This is a minor issue, but an annoyance because part of the unix standard is a linebreak at the end of the final line -- multiple checkers and source control tools add warnings or complain if that linebreak isn't there. Because PDE constantly re-introduces the error after every file load and edit, it is easy to accidentally propagate the error back into e.g. version control again and again even after removing it.

## Expected Behavior

PDE respects the final linebreak when loading a file.

## Current Behavior

PDE editor tabs load text (e.g. in Java or Python mode) with the final linebreak removed. Saving the tab also saves the altered end-of-file.

## Steps to Reproduce

1. create a PDE file with a a terminal line break
2. save the file
3. open the file in PDE -- the final linebreak does not load
4. alter the line and save -- the final linebreak has been removed.

## Your Environment

* Processing version: 3.3.6
* Operating System and OS version: macOS 10.12.6

**Comment by benfry** [comment]
Hm, we'd need to either always write one, or never. I don't particularly care either way.

I've never heard of the extra \n as part of the Unix spec; if you've got a source that'd at least give us rationale for picking the alternate. 

Note that this should also probably change inside `saveStrings()` as well.

**Comment by monkstone** [comment]
I can reproduce the same behaviour on Ubuntu linux, if I create a pde file using vim, with a line break. Then I insert a line in the code in processing ide and save, the final line break is removed.
see warning from vim editor:-
```bash
"line_break.pde" [noeol]
```

**Comment by benfry** [comment]
No need to reproduce further, we're explicitly not writing the final newline. The question is whether we always write it or never write it, and some rationale for either behavior for when the pitchforks come out because I've started adding newlines to files.

**Comment by jeremydouglass** [comment]
Here is one writeup of the canonical answer: "Because that's how the POSIX standard defines it."

-  Why should text files end with a newline? https://stackoverflow.com/a/729795/7207622
-  No newline at end of file https://stackoverflow.com/questions/5813311/no-newline-at-end-of-file

That's why tools git (and vim) -- and interfaces like SourceTree -- print a "No newline at end of file" warning.

To be clear, PDE doesn't enforce stripping the terminal linebreak at save time. Instead, it incorrectly discards one (and only one) final linebreak every time it loads, so you have to manually re-add it each time. If you forget, the file becomes invalid. So this is not about adding something new, it is about not removing something.

Here is an example: Create a text file in PDE with the number 1 followed by 5 linebreaks. Here it is with line numbering to show the blanks:

```
1. 1
2.
3.
4.

Save the file and re-open it.

1. 1
2.
3.

Make a change to the first line, save the file, and re-open it. It keeps getting shorter.

1. 12
2.

Make a change, save the file, and re-open it.

1. 123

The file now is one line long, but after another PDE save it will now technically be 0 lines long according to the POSIX standard -- because a string with no terminating linebreak isn't a full line.

**Comment by benfry** [comment]
Great, thanks. I can work with that. It's a simple fix, but touches a few places, and I want to make sure we're 100% on it before making a mess.

**Comment by benfry** [comment]
Fixed for 3.3.7 (or 3.4 if we call it that).
